devenv/lib/venv.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import os
+import shutil
+from typing import Optional
+
+from devenv import pythons
+from devenv.constants import bin_root
+from devenv.constants import venvs_root
+from devenv.lib import config
+from devenv.lib import fs
+from devenv.lib import proc
+
+VENV_OK = 1
+VENV_VERSION_MISMATCH = 2
+VENV_NOT_PRESENT = 3
+VENV_NOT_CONFIGURED = 4
+
+
+# example venv configuration section:
+#
+# [venv.sentry-kube]
+# python = 3.11.6
+# requirements = k8s/cli/requirements.txt
+# path = optional
+# editable =
+#   k8s/cli
+#   k8s/cli/libsentrykube
+# bins =
+#   sentry-kube
+#   sentry-kube-pop
+#
+# [venv.salt]
+# python = 3.10.13
+# requirements = salt/requirements.txt
+# bins =
+#   salt-ssh
+#   ...
+#
+# example usage:
+#
+# venv_dir, python_version, requirements, editable_paths, bins = get(reporoot, "sentry-kube")
+# url, sha256 = config.get_python(reporoot, python_version)
+# ensure(path, python_version, url, sha256)
+# sync(venv_dir, requirements, editable_paths, bins)
+def get(
+    reporoot: str, name: str
+) -> tuple[str, str, str, Optional[tuple[str, ...]], Optional[tuple[str, ...]]]:
+    cfg = config.get_repo(reporoot)
+
+    if not cfg.has_section(f"venv.{name}"):
+        raise KeyError(f"section venv.{name} not found in repo config")
+
+    venv = cfg[f"venv.{name}"]
+    reponame = os.path.basename(reporoot)
+    venv_dir = venv.get("path", f"{venvs_root}/{reponame}-{name}")
+    editable_paths = venv.get("editable", None)
+    if editable_paths is not None:
+        editable_paths = tuple(
+            f"{reporoot}/{path}" for path in editable_paths.strip().split("\n")
+        )
+
+    bins = venv.get("bins", None)
+    if bins is not None:
+        bins = tuple(bins.strip().split("\n"))
+
+    return (
+        venv_dir,
+        venv["python"],
+        f"{reporoot}/{venv['requirements']}",
+        editable_paths,
+        bins,
+    )
+
+
+def sync(
```

Comment:
It would be nice to have docstrings for functions to get a quick summary of what they are doing

Reply:
sure thing, i can add em in https://github.com/getsentry/devenv/pull/80